osu.Game.Rulesets.Taiko/UI/DrawableTaikoMascot.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Allocation;
+using osu.Framework.Audio.Track;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Textures;
+using osu.Game.Beatmaps.ControlPoints;
+using osu.Game.Graphics.Containers;
+
+namespace osu.Game.Rulesets.Taiko.UI
+{
+    public class DrawableTaikoMascot : BeatSyncedContainer
+    {
+        private TaikoMascotTextureAnimation idleDrawable, clearDrawable, kiaiDrawable, failDrawable;
+        private EffectControlPoint lastEffectControlPoint;
+
+        public Bindable<TaikoMascotAnimationState> PlayfieldState;
+
+        public TaikoMascotAnimationState State { get; private set; }
```

Comment:
I turned `PlayfieldState` into a private field now, while making another method called `SetPlayfieldState(state)` (therefore making this variable set only).
I also annotated the method (with a <remarks> section, mentioning the `ShowState` method) to *really* make sure, that this isn't for reading or setting the end product state.

I also read through the C# docs on `protected`, when applying your suggestion on this I noticed I couldn't use this one, because I'm accessing **`DrawableTaikoMascot`**`.State` on the parts where I have no control over whether it's the test version (`TestDrawableTaikoMascot`) or not.

I hope that should be enough for anybody else reading this code. Do you have any other thoughts on this?

Reply:
The solution you're describing kind of sounds like a Java-style method setter instead of an actual C# property setter. I can't comment on anything else without seeing the code.

I'm also not sure why you think `protected` is unviable. You need it for reading in the test scenes, so you might as well scope down types there from `DrawableTaikoMascot` to the test-specific one, or if *really* necessary, cast (but I think it shouldn't be necessary).

Reply:
I'll write what I'm observing and thinking here:
If you take a look at [here](https://github.com/ppy/osu/pull/8857/files#diff-e27de7f82e0e1bbfbecdaf785487aaacR102), you see that I have instances of the drawable from the `mascots` list (where they are being added [here](https://github.com/ppy/osu/pull/8857/files#diff-e27de7f82e0e1bbfbecdaf785487aaacR196)).

Now the problem is that the ~~Test~~DrawableTaikoMascot (in this case) is being provided from [TaikoLegacySkinTransformer](https://github.com/ppy/osu/pull/8857/files#diff-64da4d88b2512d381da640cd4b621b6cR88) (then given over to SkinnableDrawable then over to TaikoPlayfield and then finally to my tests and osu!taiko).

So in order to get a testing variant of the drawable with the easiest way possible (that I can come up with right now) I would have to add another value to [TaikoSkinComponents](https://github.com/ppy/osu/pull/8857/files#diff-052d44fe92731567f89e1c39414d6d53R18), and handle that also in TaikoLegacySkinTransformer.

Reply:
The `.ChildrenOfType<T>()` extension method should come in useful here. Check out other usages of that method in test code.

Reply:
In what way? I'm already using it in `collectMascots()` and `collectPlayfields()`.

Reply:
```diff
-                var mascot = playfield.ChildrenOfType<DrawableTaikoMascot>()
-                                      .SingleOrDefault();
+                var mascot = playfield.ChildrenOfType<TestDrawableTaikoMascot>()
+                                      .SingleOrDefault();
```
no? (Leaving aside the fact that you could probably entirely skip iterating over playfields.)

But fine, if you insist, I can do a pass over this myself in the coming days.